### ISSUE TITLE
assets.json contains duplicate URLs

### ISSUE THREAD

**Issue by emanruse** [uBlockOrigin/uBlock-issues#3099]
### Prerequisites

- [X] I verified that this is not a filter list issue. Report any issues with filter lists or broken website functionality in the [uAssets issue tracker](https://github.com/uBlockOrigin/uAssets/issues).
- [X] This is **NOT** a [_YouTube_](https://www.github.com/uBlockOrigin/uAssets/issues/20586), [_Facebook_](https://www.github.com/uBlockOrigin/uAssets/issues/3367) or [_Twitch_](https://www.github.com/uBlockOrigin/uAssets/issues/5184) report. _These sites MUST be reported by clicking their **respective links**_.
- [X] This is not a support issue or a question. For support, questions, or help, visit [/r/uBlockOrigin](https://www.reddit.com/r/uBlockOrigin/).
- [X] I performed a [cursory search of the issue tracker](https://github.com/uBlockOrigin/uBlock-issues/issues?q=is%3Aissue) to avoid opening a duplicate issue.
- [X] The issue is not present after disabling uBO in the browser.
- [X] I checked the [documentation](https://github.com/gorhill/uBlock/wiki) to understand that the issue I am reporting is not normal behavior.

### I tried to reproduce the issue when...

- [X] uBO is the only extension.
- [X] uBO uses default lists and settings.
- [X] using a new, unmodified browser profile.

### Description

`assets/assets.json` contains duplicate URLs.

### A specific URL where the issue occurs.

```yaml
https://raw.githubusercontent.com/gorhill/uBlock/master/assets/assets.json
```

### Steps to Reproduce

**Disclaimer**: I am not reporting is as a filter list issue, as it is directly related to a file contained in this repo, i.e. part of uBO's code. I noticed the issue while trying to create my own script, which caches the assets locally, so that multiple browser profiles can use them from the LAN.

----

The master file, which organizes the assets, contains duplicate URLs.

### Duplication type 1: Same filter, same section

Example:

`IRN-0` in `contentURL` section has this URL twice:

```
https://raw.githubusercontent.com/MasterKia/PersianBlocker/main/PersianBlocker.txt
```

I also don't know if it is normal the `contentURL` should contain more than one URL at all, i.e. if the others should be in section `cdnURLs`. The name (contentURL vs contentURL**s**) implies one should contain only one URL.

### Duplication type 2: Same filter, different section

In the same JSON file, there are also URLs which are contained both in `contentURL` and in `cdnURLs`, thus also resulting in duplication. Example:

Again in `IRN-0` this URL is duplicated:

```
https://cdn.statically.io/gh/MasterKia/PersianBlocker/main/PersianBlocker.txt
```

In `urlhaus-1` this one is duplicated:

```
https://malware-filter.gitlab.io/urlhaus-filter/urlhaus-filter-ag-online.txt
```

### Duplication type 3: Different filter lists

There is an even more strange issue - this URL:

```
https://cdn.jsdelivr.net/gh/uBlockOrigin/uAssetsCDN@main/filters/annoyances-cookies.txt
```

is contained *in 2 different lists*: `ublock-cookies-easylist` and `ublock-cookies-adguard`, which means that disabling one of the lists would not actually disable that list. Both must be disabled for this to be effective, which is probably not what a user would expect. IOW, even if the deduplication algorithm of uBO would neutralize the effect of the other duplications, cases like this one may actually have side effects.

----

The full list of duplicated URLs is:

```
$ jq -r '.[].contentURL, .[].cdnURLs' assets.json | grep -Eo 'https:[^"]+' | sort | uniq -cd
      2 https://cdn.jsdelivr.net/gh/uBlockOrigin/uAssetsCDN@main/filters/annoyances-cookies.txt
      2 https://cdn.statically.io/gh/MasterKia/PersianBlocker/main/PersianBlocker.txt
      2 https://cdn.statically.io/gh/uBlockOrigin/uAssetsCDN/main/filters/annoyances-cookies.txt
      2 https://malware-filter.gitlab.io/phishing-filter/phishing-filter.txt
      2 https://malware-filter.gitlab.io/urlhaus-filter/urlhaus-filter-ag-online.txt
      2 https://raw.githubusercontent.com/MasterKia/PersianBlocker/main/PersianBlocker.txt
      2 https://secure.fanboy.co.nz/fanboy-cookiemonster_ubo.txt
      2 https://secure.fanboy.co.nz/fanboy-social_ubo.txt
      2 https://ublockorigin.github.io/uAssets/filters/annoyances-cookies.txt
      2 https://ublockorigin.github.io/uAssetsCDN/filters/annoyances-cookies.txt
      2 https://ublockorigin.pages.dev/filters/annoyances-cookies.txt
```

### Expected behavior

No duplicates.

### Actual behavior

Duplicates.

### uBO version

1.55.0

### Browser name and version

Firefox 121.0.1 (64-bit)

### Operating System and version

Fedora 38

**Comment by gorhill** [comment]
All this is by design -- I will take no action. In case you feel like asking "Please explain", sorry I just do not have the free time to spend explaining all the gory details as to why this is the way it is.

**Comment by gwarser** [comment]
1. Looks like oversight to me. @gorhill ?
2. Original list source can be a CDN. Nothing prevents to use it for mass distribution.
3. It's a supplemental list with uBO-specific syntax and uBO "trusted" filters. Your line of thought seems inverted to me - you should not enable filter lists you don't need.

**Comment by gorhill** [comment]
> 1. Looks like oversight to me

There are multiple `contentURL` for Iran because sometimes GitHub was blocked in thecountry, there is an issue somewhere where this was discussed, we needed fallback URLs in case GitHub becomes inaccessible in Iran.

**Comment by gorhill** [comment]
> for Iran, 2 out of 3 elements are duplicates

Right, an oversight.

Related commit/issue:
- https://github.com/gorhill/uBlock/commit/103bdb63c5
- https://github.com/uBlockOrigin/uAssets/discussions/17921

**Comment by gorhill** [comment]
> I am not asking "Please explain"

This is also related to this other issue:

> wonder why uBO does not export settings as per the schema you are showing, so that one doesn't need to handcraft it for usage in policies.json.  Is that something to be done or is there another reason?

uBO's code base is almost 10 years old and a lot of the choices made in a distant past still need to be carried to the present, I would have to spend a lot of time researching the commit history to come up with long explanations about why things are the way they are, and I just do not have the time to go through that sort of work. I just want to focus on short, to-the-point actionable issues that need to be fixed.

For point 1, I ended up missing the actual real duplication issue because of that last sentence that asked me to explain:

> I also don't know if it is normal the `contentURL` should contain more than one URL at all, i.e. if the others should be in section `cdnURLs`. The name (`contentURL` vs `contentURL`s) implies one should contain only one URL.

If you look at commit history using `git` tools, probably you would find out originally `contentURL` was meant to be a single URL. Things changed and eventually I had to support possibility of more than one URL, hence why not-plural `contentURL` is named the why it is -- changing the property name would complicate backward compatibility.

And so on. That sort of questions should be brought to a discussion item instead of an issue item, as I really just want to focus on on actionable issues, while with discussion items I or others can contribute and do not require changes to the code.
